src/vm/functions/mod.rs

```diff
@@ -397,3 +408,113 @@ fn special_contract_of(args: &[SymbolicExpression], env: &mut Environment, conte
     let contract_principal = Value::Principal(PrincipalData::Contract(contract_identifier.clone()));
     Ok(contract_principal)
 }
+
+fn special_principal_of(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    // (principal-of (..))
+    // arg0 => (buff 33)
+    check_argument_count(1, args)?;
+
+    runtime_cost!(cost_functions::PRINCIPAL_OF, env, 0)?;
+
+    let param0 = eval(&args[0], env, context)?;
+    let pub_key = match param0 {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => {
+            if data.len() != 33 {
+                return Err(CheckErrors::TypeValueError(BUFF_33, param0).into())
+            }
+            data
+        },
+        _ => return Err(CheckErrors::TypeValueError(BUFF_33, param0).into())
+    };
+
+    let addr = StacksAddress::from_public_keys(C32_ADDRESS_VERSION_TESTNET_SINGLESIG, &AddressHashMode::SerializeP2PKH, 1, &vec![Secp256k1PublicKey::from_slice(&pub_key).unwrap()]).unwrap();
+    let principal = addr.to_account_principal();
+
+    Ok(Value::Principal(principal))
+}
+
+fn special_secp256k1_recover(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    // (principal-of (..))
+    // arg0 => (buff 32), arg1 => (buff 65)
+    check_argument_count(2, args)?;
+
+    runtime_cost!(cost_functions::SECP256K1RECOVER, env, 0)?;
+
+    let param0 = eval(&args[0], env, context)?;
+    let message = match param0 {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => {
+            if data.len() != 32 {
+                return Err(CheckErrors::TypeValueError(BUFF_32, param0).into())
+            }
+            data
+        },
+        _ => return Err(CheckErrors::TypeValueError(BUFF_32, param0).into())
+    };
+
+    let param1 = eval(&args[1], env, context)?;
+    let signature = match param1 {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => {
+            if data.len() != 65 {
+                return Err(CheckErrors::TypeValueError(BUFF_65, param1).into())
+            }
+            if data[64] > 3 {
+                return Err(CheckErrors::InvalidSecp65k1Signature.into())
```

Comment:
I think this should also be a handle-able error code in Clarity (i.e., return `Value::err_uint(2)`), and for the data.len() check, it should only be a TypeValueError if `data.len() > 65` (that's all that the type system will enforce), for `data.len() < 65`, we should probably have `Value::err_uint(2)` as well.

Reply:
sure, I'll make the change.  The `verify` function should behave the way as well, then.

Reply:
address in 3907ed5, except I changed verify to return `false` instead of throwing an error